### PR TITLE
feat: feedback system backend

### PR DESCRIPTION
**Pull request by SrIzan10**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,19 +2,27 @@
   "name": "automata",
   "version": "1.0.0",
   "scripts": {
-    "dev": "tsc-watch --onSuccess \"node dist/index.js\"",
+    "dev": "pocketbase-typegen --env --out ./util/pbtypes.ts;tsc-watch --onSuccess \"node dist/index.js --dev\"",
     "build": "tsc"
   },
   "type": "module",
   "dependencies": {
+    "axios": "^1.5.0",
```
Comment (jacoobes) [inline package.json]: switch to using vanilla fetch, which is native in version 18

File: index.ts [hunk 2]
```diff
@@ -36,6 +59,76 @@ app.post('/wh/updateDocsJson', async (req, res) => {
...
+app.post('/tutorial/feedback', feedbackRateLimit, async (req, res) => {
+	const body = JSON.parse(JSON.stringify(req.body)) as FeedbackRequestBody
+	// validation of request body
+	if (
+		typeof body.turnstileToken !== "string" ||
+		// type of string should be "up" or "down" but it's checked later
+		typeof body.feedback !== "string" ||
+		typeof body.route !== "string"
+  	)
+    return res
```
Comment (jacoobes) [inline index.ts:69]: formatted a little weird here

File: index.ts [hunk 3]
```diff
@@ -36,6 +59,76 @@ app.post('/wh/updateDocsJson', async (req, res) => {
...
+	// actual database recording
+	const data: FeedbackRecord = { ...body }
+	await pb.collection('feedback').create(data)
+	res.send({
+		successful: true,
+		message: "Feedback recorded!",
+	});
+
+	// webhook
+	const webhook = new Webhook(new URL(process.env.DEV_WEBHOOK!), 'sern Guide Feedback (by automata)', 'https://avatars.githubusercontent.com/u/129876409?v=4')
```
Comment (jacoobes) [inline index.ts]: sern Guide Feedback -> Guide Feedback

File: package.json [hunk 4]
```diff
@@ -2,19 +2,27 @@
...
     "body-parser": "^1.20.2",
+    "cors": "^2.8.5",
     "dotenv": "^16.0.3",
     "execa": "^7.1.1",
-    "express": "^4.18.2"
+    "express": "^4.18.2",
+    "express-rate-limit": "^6.11.1",
+    "form-data": "^4.0.0",
```
Comment (jacoobes) [inline package.json]: remove in favor of native URLSearchParams

File: index.ts [hunk 5]
```diff
@@ -1,12 +1,35 @@
...
-import bodyParser from 'body-parser';
 import 'dotenv/config';
 import { execa } from 'execa';
 import { validateJsonWebhook } from './util/validateJsonWebhook.js';
 import babashkaScripts from './babashka/scripts.json' assert { type: 'json' };
+import PocketBase from 'pocketbase';
+import { FeedbackRecord } from './util/pbtypes.js';
+import { FeedbackRequestBody } from './util/types.js';
+import axios from 'axios';
+import FormData from 'form-data';
```
Comment (jacoobes) [inline index.ts]: we can use URLSearchParams instead of form-data

**Review by jacoobes — CHANGES_REQUESTED** [review]

File: index.ts [hunk 6]
```diff
@@ -36,6 +59,76 @@ app.post('/wh/updateDocsJson', async (req, res) => {
...
+			successful: false,
+			error: "You have something missing in your request!",
+		});
+	if (body.feedback !== "up" && body.feedback !== "down")
+		return res
+			.status(400)
+			.send({
+				successful: false,
+				error: "Feedback must be either 'up' or 'down'!",
+			});
```
Comment (Murtatrxx) [inline index.ts:81]: Why check like that? Just check up & firstly

File: index.ts (hunk 2 above)
Comment (SrIzan10) [reply index.ts:69]: yeah, big L for prettier

File: package.json (hunk 4 above)
Comment (SrIzan10) [reply package.json]: form-data you mean?

File: index.ts (hunk 5 above)
Comment (SrIzan10) [reply index.ts]: Axios needs form-data to work so we should keep it

File: package.json (hunk 1 above)
Comment (SrIzan10) [reply package.json]: even though I like fetch, axios is simpler imo so I'd keep it

File: index.ts (hunk 6 above)
Comment (SrIzan10) [reply index.ts:81]: the request body can be tampered by the user, so we check the types to not get any javascript errors that could crash the bot

File: index.ts (hunk 2 above)
Comment (SrIzan10) [reply index.ts:69]: it looks good on my part:

File: util/pbtypes.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+/**
```
Comment (jacoobes) [inline util/pbtypes.ts:1]: should we include this file or should this be .gitignored?

Comment (jacoobes) [reply util/pbtypes.ts:1]: since its generated

Comment (SrIzan10) [reply util/pbtypes.ts:1]: it's generated but it shouldn't get gitignored because other devs normally don't have access to the database

Comment (SrIzan10) [reply util/pbtypes.ts:1]: there's also no sensitive info in the file

**Review by jacoobes — APPROVED** [review]

**Pull request merged**